gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java

```diff
@@ -0,0 +1,165 @@
+   *
+   * @param name the name of the {@link TaskScheduler}
+   */
+  @Synchronized
+  final void start(Optional<String> name) {
+    if (!this.isStarted) {
+      this.startImpl(name);
+      this.isStarted = true;
+    }
+  }
+
+  /**
+   * Start the {@link TaskScheduler}.
+   *
+   * @param name the name of the {@link TaskScheduler}
+   */
+  abstract void startImpl(Optional<String> name);
+
+  /**
+   * Schedules a subclass of {@link ScheduledTask} to run periodically.
+   *
+   * @param task the subclass of {@link ScheduledTask} to run every period
+   * @param period the period between successive executions of the task
+   * @param unit the time unit of the initialDelay and period parameters
+   */
+  public final void schedule(final T task, final long period, final TimeUnit unit) throws IOException {
+    Preconditions.checkArgument(isStarted, "TaskScheduler is not started");
+    try {
+      CancellableTask<K, T> cancellableTask = this.cancellableTaskMap.get(task.getKey(),
+              new Callable<CancellableTask<K, T>>() {
+                @Override
+                public CancellableTask<K, T> call() {
+                  return scheduleImpl(task, period, unit);
+                }
+              });
+      if (cancellableTask.getScheduledTask() != task) {
+        throw new IOException("Failed to schedule task with duplicate key");
+      }
+    } catch (ExecutionException e) {
+      throw new IOException("Failed to schedule task", e);
+    }
+  }
+
+  /**
+   * Schedules a subclass of {@link ScheduledTask} to run periodically.
+   *
+   * @param task the subclass of {@link ScheduledTask} to run every period
+   * @param period the period between successive executions of the task
+   * @param unit the time unit of the period parameter
+   */
+  abstract CancellableTask<K, T> scheduleImpl(T task, long period, TimeUnit unit);
+
+  /**
+   * Gets all {@link ScheduledTask}s.
+   *
+   * @return the {@link ScheduledTask}s
+   */
+  public final Iterable<T> getScheduledTasks() {
+    return Iterables.transform(this.cancellableTaskMap.asMap().values(),
+          new Function<CancellableTask<K, T>, T>() {
+            @Override
+            public T apply(CancellableTask<K, T> cancellableTask) {
+              return cancellableTask.getScheduledTask();
+            }
+          });
+  }
+
+  /**
+   * Gets the {@link ScheduledTask} with the specified key.
+   *
+   * @return the {@link ScheduledTask}
+   */
+  public final Optional<T> getScheduledTask(K key) {
+    CancellableTask<K, T> cancellableTask = this.cancellableTaskMap.getIfPresent(key);
+    if (cancellableTask != null) {
+      return Optional.of(cancellableTask.getScheduledTask());
+    }
+    return Optional.absent();
+  }
+
+  /**
+   * Attempts to cancel the specified {@link ScheduledTask}.
+   *
+   * @param task the {@link ScheduledTask} to cancel
+   * @return true if the {@link ScheduledTask} was canceled; otherwise, false
+   */
+  public boolean cancel(T task) {
```

Comment:
Well, all implementations of the cancel method are currently safe.  For instance the `HashedWheelTimer` `Timeout` states: "If the task has been executed or cancelled already, it will return with no side effect."  Additionally, the `Future` states the `cancel` method returns: "false if the task could not be cancelled, typically because it has already completed normally".  I don't think it is unreasonable for interfaces to require that the implementations are thread-safe.  This is done in many other places.  My preference is to reduce high level locking as much as possible.